cmd/istioctl/istioctl_test.go

```diff
@@ -15,159 +15,359 @@
+			return nil, err
+		}
+		res = append(res, config)
 	}
+	return res, nil
 }
 
-func TestGet(t *testing.T) {
-	rootSetup(t)
-	file = "testdata/dest-policy.yaml"
-	if err := postCmd.RunE(postCmd, []string{}); err != nil {
-		t.Fatalf("Could not create destination policy: %v", err)
+func (st *StubClient) setupTwoRouteRuleMap() {
+	st.KeyConfigMap = make(map[model.Key]apiserver.Config)
+	key1 := model.Key{
+		Name:      "test-v1",
+		Namespace: namespace,
+		Kind:      "route-rule",
 	}
-
-	file = ""
-	if err := getCmd.RunE(getCmd, []string{"destination-policy", "world-cb"}); err != nil {
-		t.Fatalf("Did not find world-cb %v", err)
+	key2 := model.Key{
+		Name:      "test-v2",
+		Namespace: namespace,
+		Kind:      "route-rule",
 	}
-
-	if err := deleteCmd.RunE(deleteCmd, []string{"destination-policy", "world-cb"}); err != nil {
-		t.Fatalf("Could not delete world-cb: %v", err)
+	st.KeyConfigMap[key1] = apiserver.Config{
+		Name: "test-v1",
+		Type: "route-rule",
 	}
-
-	if err := getCmd.RunE(getCmd, []string{"destination-policy", "world-cb"}); err == nil {
-		t.Fatalf("Found world-cb after deletion ")
+	st.KeyConfigMap[key2] = apiserver.Config{
+		Name: "test-v2",
+		Type: "route-rule",
 	}
 }
 
-func TestNewGet(t *testing.T) {
-	rootSetup(t)
-
-	file = "testdata/dest-policy.yaml"
-	if err := postCmd.RunE(postCmd, []string{}); err != nil {
-		t.Fatalf("Could not create destination policy: %v", err)
-	}
+func (st *StubClient) setupDeleteKeys() {
+	st.WantKeys = make(map[model.Key]struct{})
+	st.WantKeys[model.Key{Name: "test-v1", Namespace: "default", Kind: "route-rule"}] = struct{}{}
+	st.WantKeys[model.Key{Name: "test-v2", Namespace: "default", Kind: "route-rule"}] = struct{}{}
+}
 
-	file = ""
-	if err := getCmd.RunE(getCmd, []string{"route-rule"}); err != nil {
-		t.Fatalf("Could not list routes: %v", err)
+func (st *StubClient) verifyKeyConfig(key model.Key, config apiserver.Config) error {
+	wantConfig, ok := st.KeyConfigMap[key]
+	if !ok {
+		return fmt.Errorf("received unexpected key/config pair\n key: %+v\nconfig: %+v", key, config)
 	}
-
-	// Plural form
-	if err := getCmd.RunE(getCmd, []string{"route-rules"}); err != nil {
-		t.Fatalf("Could not list routes: %v", err)
+	// ToDo: test spec as well
+	if strings.Compare(wantConfig.Name, config.Name) != 0 {
+		return fmt.Errorf("received unexpected config name: %s, wanted: %s", config.Name, wantConfig.Name)
 	}
-
-	// Short output
-	outputFormat = "short"
-	if err := getCmd.RunE(getCmd, []string{"route-rules"}); err != nil {
-		t.Fatalf("Could not list routes: %v", err)
+	if strings.Compare(wantConfig.Type, config.Type) != 0 {
+		return fmt.Errorf("received unexpected config type: %s, wanted: %s", config.Type, wantConfig.Type)
 	}
+	return nil
+}
 
-	// YAML output
-	outputFormat = "yaml"
-	if err := getCmd.RunE(getCmd, []string{"route-rules"}); err != nil {
-		t.Fatalf("Could not list routes: %v", err)
+func TestClientSideValidation(t *testing.T) {
```

Comment:
I would suggest we keep all config validation at server side only. Otherwise we risk having old clients providing a (locally valid) config that is invalid at a newer server.